File: src/alfenctl/web/static/js/network.js
/* The Network tab: where the charger is, and how it got there.
 *
 * The interface addresses (read-only, the way the CLI prints them), the
 * Wi-Fi radio's scan and join, the live smart-meter wiring test, and the
 * custom Modbus register map for a meter the charger does not know.
 */

import { html, useEffect, useState } from '../vendor/preact-htm.module.js';
import {
  Loading,
  PanelCard,
  PanelError,
  usePanel,
} from './panels.js';
import { Card, Row } from './ui.js';

function WifiScan({ readOnly, busy, api, toast }) {
  const [doc, setDoc] = useState(null);
  const [scanning, setScanning] = useState(false);
  const [error, setError] = useState('');
  const [psk, setPsk] = useState('');
  const [joining, setJoining] = useState(null);
  const [enableFirst, setEnableFirst] = useState(true);

  const scan = async () => {
    setScanning(true);
    setError('');
    try {
      setDoc(await api.get('/wifi/scan', enableFirst ? { enable: '1' } : {}));
    } catch (err) {
      setError(err.message);
    } finally {
      setScanning(false);
    }
  };

  const join = (net) =>
    api
      .post('/wifi', {
        action: 'connect',
        ssid: net.ssid,
        psk: psk || null,
        security: net.security,
      })
      .then(() => toast.ok(`Asked to join ${net.ssid}.`))
      .catch((err) => toast.error(err.message))
      .finally(() => setJoining(null));

  const nets = doc?.networks || [];
  return html`<${Card} title="Wi-Fi">
    <div class="actions-row">
      <button class="btn" disabled=${scanning || busy} onClick=${scan}>
        ${scanning ? 'Scanning...' : 'Scan for networks'}
      </button>
      <label class="toggle" title="a scan with the radio off finds nothing; this switches it on first">
        <input
          type="checkbox"
          checked=${enableFirst}
          disabled=${readOnly || busy}
          onChange=${(e) => setEnableFirst(e.target.checked)}
        />
        switch the radio on first
      </label>
    </div>
    ${error && html`<p class="card-note bad">${error}</p>`}
    ${doc?.note && html`<p class="note">${doc.note}</p>`}
    ${nets.length > 0 &&
    /* A network is a name, a line about it, and one button, which is the
     * shape `.entry` already draws for Alfen's releases -- and the
     * passphrase, when it is asked for, has the width of the entry rather
     * than of a table cell.  As a three-column table this card had to
     * claim two tracks of the grid, and a card two tracks wide ends the
     * row for every card after it. */
    html`<div class="scroller spaced">
      ${nets.map(
        (n) => html`<div class="entry" key=${n.ssid}>
          <div class="head">
            <span class="name data">${n.ssid}</span>
            ${!readOnly &&
            html`<span class="act">
              <button
                class="btn small"
                disabled=${busy || !n.supported || joining === n.ssid}
                title=${n.supported ? '' : 'the charger cannot join this security mode'}
                onClick=${() => {
                  setJoining(n.ssid);
                  setPsk('');
                }}
              >
                join
              </button>
            </span>`}
          </div>
          <div class="meta muted">
            ${n.securityLabel} -- ${n.signal} (${n.signalDbm} dBm)
          </div>
          ${joining === n.ssid &&
          html`<div class="actions-row">
            <input
              type="password"
              placeholder="passphrase"
              class="grow"
              value=${psk}
              onInput=${(e) => setPsk(e.target.value)}
            />
            <button class="btn small primary" disabled=${busy} onClick=${() => join(n)}>
              join
            </button>
            <button class="btn small ghost" onClick=${() => setJoining(null)}>cancel</button>
          </div>`}
        </div>`
      )}
    </div>`}
    ${!readOnly &&
    html`<div class="actions-row">
      <button
        class="btn ghost"
        disabled=${busy}
        onClick=${() =>
          api
            .post('/wifi', { action: 'enable' })
            .then(() => toast.ok('Wi-Fi radio enabled.'))
            .catch((err) => toast.error(err.message))}
      >
        Enable radio
      </button>
      <button
        class="btn ghost"
        disabled=${busy}
        onClick=${() =>
          api
            .post('/wifi', { action: 'disconnect' })
            .then(() => toast.ok('Wi-Fi radio switched off.'))
            .catch((err) => toast.error(err.message))}
      >
        Switch radio off
      </button>
    </div>`}
  <//>`;
}

function MeterTest({ busy, api }) {
  const [doc, setDoc] = useState(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState('');

  const read = async () => {
    setLoading(true);
    setError('');
    try {
      setDoc(await api.get('/meter-test'));
    } catch (err) {
      setError(err.message);
    } finally {
      setLoading(false);
    }
  };
  useEffect(() => {
    if (doc === null && !error) read();
  }, []);
  if (doc === null && !error) {
    return html`<${Loading} loading=${loading} what="Reading the meter test..." title="Smart-meter test" />`;
  }

  const readings = doc?.readings || [];
  return html`<${Card}
    title="Smart-meter test"
    actions=${html`<button class="btn small ghost" disabled=${busy} onClick=${read}>
      refresh
    </button>`}
  >
    ${error
      ? html`<${PanelError} error=${error} loading=${loading} onRetry=${read} />`
      : readings.length === 0
        ? html`<div class="empty">
            No live meter readings. The charger reports them once the
            <code>meter4</code> category has been read -- open the Properties
            tab and read it first.
          </div>`
        : html`<div class="rows">
            ${readings.map(
              (m) => html`<${Row}
                key=${m.label}
                k=${m.label}
                v=${m.value === null ? '—' : `${m.value} ${m.unit}`}
              />`
            )}
          </div>`}
  <//>`;
}

function MeterMap({ readOnly, busy, api, toast }) {
  const [doc, loading, error, read] = usePanel('meter-map', () => api.get('/meter-map'));
  if (error) return html`<${PanelError} error=${error} loading=${loading} onRetry=${read} title="Custom meter map" />`;
  if (!doc) return html`<${Loading} loading=${loading} what="Reading the meter map..." title="Custom meter map" />`;

  const map = doc.meterMap || {};
  const entries = map.entries || [];
  return html`<${Card}
    title="Custom meter map"
    actions=${entries.length > 0 && html`<span class="badge">${entries.length} of ${map.capacity}</span>`}
    width="full"
  >
    ${entries.length === 0
      ? html`<div class="empty">
          No custom register map is installed; the charger is reading a meter
          it knows, or none at all.
        </div>`
      : html`<div class="table-wrap">
          <table>
            <thead>
              <tr>
                <th>Measurand</th>
                <th>Register</th>
                <th>Type</th>
                <th>Scale</th>
              </tr>
            </thead>
            <tbody>
              ${entries.map(
                (e, i) => html`<tr key=${i}>
                  <td>${e.measurand}</td>
                  <td class="data">${e.register}</td>
                  <td>${e.dataType}</td>
                  <td>${e.factor}</td>
                </tr>`
              )}
            </tbody>
          </table>
        </div>`}
    ${!readOnly &&
    html`<div class="actions-row">
      <input
        type="file"
        accept=".json,application/json"
        disabled=${busy}
        onChange=${(e) => {
          const file = e.target.files?.[0];
          if (!file) return;
          api
            .upload('/meter-map', file)
            .then(() => {
              toast.ok('Register map written. Check it with the meter test.');
              read();
            })
            .catch((err) => toast.error(err.message));
          e.target.value = '';
        }}
      />
      <span class="note">one of Alfen's register-map JSON files</span>
    </div>`}
  <//>`;
}

export function Network({ api, readOnly, busy, toast }) {
  const [netDoc, netLoading, netError, netRead] = usePanel('network', () => api.get('/network'));
  return html`<div class="grid">
    ${netError
      ? html`<${PanelError} error=${netError} loading=${netLoading} onRetry=${netRead} title="Interfaces" />`
      : netDoc
        ? html`<${PanelCard}
            title="Interfaces"
            rows=${netDoc.network?.rows}
            note="Where the charger is on the network."
            more="Read only, and as the CLI reports it: these are the addresses the
              charger answers with. The Wi-Fi radio is joined from the card beside this
              one, and a wired address is set on the charger itself."
          />`
        : html`<${Loading} loading=${netLoading} what="Reading the network..." title="Interfaces" />`}
    <${WifiScan} readOnly=${readOnly} busy=${busy} api=${api} toast=${toast} />
    <${MeterTest} busy=${busy} api=${api} />
    <${MeterMap} readOnly=${readOnly} busy=${busy} api=${api} toast=${toast} />
  </div>`;
}

File: src/alfenctl/cli/commands/network.py
"""``alfenctl network`` and ``alfenctl wifi`` -- where the charger is reachable.

``wifi`` used to do one thing, scan, and it still does when it is given no
action.  What it could not do was act on what it found: the scan printed three
``alfenctl set`` lines for you to copy.  ``wifi connect`` writes those three
registers itself, with the SSID and key length checked first.

``network show`` is the read side for every interface at once -- Ethernet,
Wi-Fi and the modem -- which is the question ``info`` does not answer.

A scan is the charger's own radio listening, so ``wifi scan`` reads the radio
first and says when the empty list it is about to print means "the radio is
off" rather than "there is nothing out there".  ``wifi scan --enable`` (or
``wifi enable``) switches it on and waits for it to come up.

**And it asks more than once.**  ``/api/wifiscan`` comes back in about three
seconds where the vendor's client allows eleven, which is the shape of an
endpoint that hands back the survey the radio has rather than waiting for a
fresh one.  A radio that has just been switched on has none yet, so an empty
list is retried a few times before it is believed.
"""

from __future__ import annotations

import argparse
import json
import sys
import time

from alfenctl import network, wifi
from alfenctl.charger import AlfenCharger
from alfenctl.cli.command import Command
from alfenctl.cli.exits import EXIT_ERROR, EXIT_OK
from alfenctl.cli.output import print_rows


def cmd_network(charger: AlfenCharger, args: argparse.Namespace) -> int:
    """Show every interface's address and state."""
    state = network.read(charger)
    rows = state.rows()
    if not rows:
        print("The charger reported none of the network properties.", file=sys.stderr)
        return EXIT_OK
    print_rows("Network", rows)
    return EXIT_OK


def _ready_radio(charger: AlfenCharger, args: argparse.Namespace) -> str | None:
    """Make sure the radio can scan, switching it on when asked to.

    Returns the reason a scan will find nothing, or None when it should
    work.  Without ``--enable`` nothing is written: a charger reached over
    Ethernet has its Wi-Fi off for a reason often enough.
    """
    state = network.read(charger)
    obstacle = state.scan_obstacle()
    if obstacle is None or state.wifi_hardware is False or not args.enable:
        return obstacle
    print("Switching the Wi-Fi radio on...", file=sys.stderr)
    network.enable(charger)
    state = network.wait_for_radio(charger)
    if not state.radio_ready:
        print(
            "warning: the radio is not running yet "
            f"({network.WIFI_STATUSES.get(state.wifi_status or -1, 'unknown')}); "
            "scanning anyway",
            file=sys.stderr,
        )
    return state.scan_obstacle()


# The endpoint answers in about three seconds, well inside the eleven the
# vendor's client allows, so it reports the survey the radio already has
# rather than running one to order.  A radio that has just come up has none.
SCAN_ATTEMPTS = 3
SCAN_RETRY_INTERVAL_S = 4.0


def _scan_repeatedly(charger: AlfenCharger, attempts: int) -> wifi.ScanReply:
    """Scan until something is heard, or the attempts run out.

    Returns the last reply either way -- including its complaints, which are
    the difference between "nothing is out there" and "that was not a scan
    result at all".
    """
    reply = wifi.ScanReply()
    for attempt in range(1, max(1, attempts) + 1):
        reply = wifi.parse_reply(charger.wifi_scan())
        if reply.networks or attempt == attempts:
            return reply
        print(
            f"Nothing heard yet; scanning again ({attempt + 1}/{attempts})...",
            file=sys.stderr,
        )
        time.sleep(SCAN_RETRY_INTERVAL_S)
    return reply


def _explain_empty(reply: wifi.ScanReply, obstacle: str | None) -> None:
    """Say why the list is empty, as precisely as the reply allows."""
    if obstacle:
        # The endpoint answered, and answered nothing, because the radio
        # doing the listening is not on.  That is not an empty band.
        print(
            f"note: {obstacle}.\n"
            "A scan is the charger's own radio listening, so it finds nothing "
            "while the radio is off.\n"
            "Switch it on and scan again with:\n"
            "  alfenctl wifi scan --enable",
            file=sys.stderr,
        )
        return
    detail = reply.summary()
    if reply.complaint:
        print(
            f"note: {detail}.\n"
            "That is not an empty band but an answer this parser did not "
            "recognise; `--debug` prints it verbatim.",
            file=sys.stderr,
        )
        return
    if detail:
        print(f"note: {detail}.", file=sys.stderr)
    print(
        "The radio is on and heard nothing. What is left to check:\n"
        "  - the band: a router broadcasting that SSID only on 5 GHz is "
        "inaudible to a 2.4 GHz radio\n"
        "  - the distance: the charger's antenna is inside a metal case, "
        "outdoors, often behind a wall\n"
        "  - the moment: try `alfenctl wifi scan` again in a minute, or after "
        "`alfenctl reboot` if the radio was only just switched on",
        file=sys.stderr,
    )


def _scan(charger: AlfenCharger, args: argparse.Namespace) -> int:
    """Scan for nearby Wi-Fi networks (the app's 'Scan Wi-Fi networks' button)."""
    obstacle = _ready_radio(charger, args)
    reply = _scan_repeatedly(charger, args.attempts)
    networks = reply.networks
    if not networks:
        _explain_empty(reply, obstacle)
    elif (detail := reply.summary()) is not None:
        print(f"note: {detail}.", file=sys.stderr)
    if args.json:
        print(
            json.dumps(
                [
                    {
                        "ssid": n.ssid,
                        "signal_dbm": n.signal_dbm,
                        "signal": n.signal_label,
                        "security": n.security_label,
                        "supported": n.supported,
                    }
                    for n in networks
                ],
                indent=2,
            )
        )
        return EXIT_OK
    if not networks:
        print("No Wi-Fi networks found.")
        return EXIT_OK
    width = max(len(n.ssid) for n in networks)
    print(f"{'SSID'.ljust(width)}  SIGNAL       SECURITY")
    for n in networks:
        sec = n.security_label if n.supported else f"{n.security_label} (unsupported)"
        print(
            f"{n.ssid.ljust(width)}  {n.signal_dbm:>4} dBm {n.signal_label:<9}  {sec}"
        )
    print(
        f"\n{len(networks)} network(s). Join one with:\n"
        f"  alfenctl wifi connect '{networks[0].ssid}' --psk <passphrase>"
    )
    return EXIT_OK


def _report(state: network.Network) -> None:
    """Print the Wi-Fi half of a reading, after a change to it."""
    rows = [(label, value) for label, value in state.rows() if "Ethernet" not in label]
    if rows:
        print_rows("Wi-Fi", rows)


def cmd_wifi(charger: AlfenCharger, args: argparse.Namespace) -> int:
    """Scan for Wi-Fi networks, join one, or leave."""
    if args.action == "scan":
        return _scan(charger, args)
    if args.action == "connect":
        password = args.psk
        if password is None and not args.open:
            import getpass

            password = getpass.getpass(f"Password for {args.ssid}: ")
        state = network.connect(
            charger, args.ssid, None if args.open else password, security=args.security
        )
        # The charger takes a moment to associate, so what it reports here is
        # the settings it accepted rather than the outcome; say so.
        print(f"Sent the settings for {args.ssid!r}.\n")
        _report(state)
        print(
            "\nThe radio associates in the background; `alfenctl network show` "
            "says whether it got there.",
            file=sys.stderr,
        )
        return EXIT_OK
    if args.action == "enable":
        if network.read(charger).wifi_hardware is False:
            print(
                "error: this charger reports no Wi-Fi radio (0x328F_0 is 0)",
                file=sys.stderr,
            )
            return EXIT_ERROR
        network.enable(charger)
        state = network.wait_for_radio(charger)
        _report(state)
        if not state.radio_ready:
            print(
                "\nThe radio has not reported itself running yet; "
                "`alfenctl network` says when it does.",
                file=sys.stderr,
            )
        return EXIT_OK
    if args.action == "disconnect":
        _report(network.disconnect(charger))
        return EXIT_OK
    if args.action == "ap":
        if args.enable is None and args.start is None:
            print("error: name --enable or --start", file=sys.stderr)
            return EXIT_ERROR
        _report(
            network.set_access_point(
                charger,
                enabled=None if args.enable is None else args.enable == "on",
                start=None if args.start is None else args.start == "on",
            )
        )
        return EXIT_OK
    return EXIT_OK


def add_parsers(
    sub: argparse._SubParsersAction, common: argparse.ArgumentParser
) -> None:
    """Add this group's commands to the root parser."""
    sub.add_parser(
        "network",
        help="show the charger's Ethernet, Wi-Fi and modem addresses",
        parents=[common],
        description="Show where this charger is on the network. The addresses "
        "are read-only here: writing an interface's own address is how a "
        "station is lost, so it stays an explicit `alfenctl set 207D_2 ...`.",
    )

    sp = sub.add_parser(
        "wifi",
        help="scan for Wi-Fi networks, or join one",
        description="Scan for the networks the charger's radio can see, or "
        "join one by writing wifiSSID (0x328A_0), wifiPSK (0x328B_0) and "
        "sysWifiEnabled (0x3284_0) together. A scan asks the charger's own "
        "radio, so it finds nothing while sysWifiEnabled is 0. With no "
        "ACTION, scans.",
    )
    wsub = sp.add_subparsers(dest="action", metavar="ACTION", required=True)
    wc = wsub.add_parser("scan", help="list the networks in range", parents=[common])
    wc.add_argument("--json", action="store_true", help="JSON instead of a table")
    wc.add_argument(
        "--enable",
        action="store_true",
        help="switch the radio on first, and wait for it: a scan is the "
        "charger's own radio listening, and finds nothing while it is off",
    )
    wc.add_argument(
        "--attempts",
        type=int,
        default=SCAN_ATTEMPTS,
        metavar="N",
        help=f"how many times to ask before believing an empty list "
        f"(default {SCAN_ATTEMPTS}); the charger reports the survey its radio "
        "already has, which just after a boot is none",
    )
    wc = wsub.add_parser("connect", help="join a network", parents=[common])
    wc.add_argument("ssid", metavar="SSID", help="the network to join")
    # Not --password: the connection options this parser inherits already have
    # one, for the charger's own login.  wifiPSK is the vendor's own name.
    wc.add_argument("--psk", help="its passphrase (prompted for when not given)")
    wc.add_argument(
        "--open", action="store_true", help="the network has no password at all"
    )
    wc.add_argument(
        "--security",
        type=int,
        metavar="N",
        help="the security type, when the default (WPA2-AES) is wrong: "
        + ", ".join(f"{v} ({n})" for v, n in sorted(network.SECURITY_TYPES.items())),
    )
    wc = wsub.add_parser(
        "enable",
        help="switch the radio on, keeping the settings",
        parents=[common],
    )
    wc = wsub.add_parser(
        "disconnect",
        help="switch the radio off, keeping the settings",
        parents=[common],
    )
    wc = wsub.add_parser("ap", help="the charger's own access point", parents=[common])
    wc.add_argument(
        "--enable", choices=("on", "off"), help="start the access point after a boot"
    )
    wc.add_argument("--start", choices=("on", "off"), help="start it now")


COMMANDS: dict[str, Command] = {
    "network": Command(cmd_network),
    "wifi": Command(cmd_wifi),
}
